src/services/doppler-billing-api-client.ts

```diff
@@ -0,0 +1,106 @@
+import { ResultWithoutExpectedErrors } from '../doppler-types';
+import { AxiosInstance, AxiosStatic } from 'axios';
+import { AppSession } from './app-session';
+import { RefObject } from 'react';
+
+export interface DopplerBillingApiClient {
+  getInvoices(page: number, pageSize: number): Promise<ResultWithoutExpectedErrors<Invoices>>;
+}
+
+interface DopplerBillingApiConnectionData {
+  jwtToken: string;
+  idUser: number;
+}
+
+export interface Invoices {
+  items: Invoice[];
+  totalItems: number;
+}
+
+export interface Invoice {
+  accountId: string;
+  product: string;
+  date: Date;
+  currency: string;
+  amount: number;
+  filename: string;
+}
+
+export class HttpDopplerBillingApiClient implements DopplerBillingApiClient {
+  private readonly axios: AxiosInstance;
+  private readonly baseUrl: string;
+  private readonly connectionDataRef: RefObject<AppSession>;
+
+  constructor({
+    axiosStatic,
+    baseUrl,
+    connectionDataRef,
+  }: {
+    axiosStatic: AxiosStatic;
+    baseUrl: string;
+    connectionDataRef: RefObject<AppSession>;
+  }) {
+    this.baseUrl = baseUrl;
+    this.axios = axiosStatic.create({
+      baseURL: this.baseUrl,
+    });
+    this.connectionDataRef = connectionDataRef;
+  }
+
+  private getDopplerBillingApiConnectionData(): DopplerBillingApiConnectionData {
+    const connectionData = this.connectionDataRef.current;
+    if (
+      !connectionData ||
+      connectionData.status !== 'authenticated' ||
+      !connectionData.jwtToken ||
+      !connectionData.userData
+    ) {
+      throw new Error('Doppler Billing API connection data is not available');
+    }
+    return {
+      jwtToken: connectionData.jwtToken,
+      idUser: connectionData.userData.user.idUser,
+    };
+  }
+
+  private mapInvoices(data: any): Invoice[] {
+    return data.map((x: any) => ({
+      accountId: x.accountId,
+      product: x.product,
+      date: x.date,
+      currency: x.currency,
+      amount: x.amount,
+      filename: x.filename,
+    }));
+  }
+
+  public async getInvoices(
+    page: number,
+    pageSize: number,
+  ): Promise<ResultWithoutExpectedErrors<Invoices>> {
+    try {
+      const { idUser, jwtToken } = this.getDopplerBillingApiConnectionData();
+      const account = 'doppler';
+
+      const response = await this.axios.request({
+        method: 'GET',
+        url: `/accounts/${account}/${idUser}/invoices?page=${page}&pageSize=${pageSize}`,
+        headers: { Authorization: `bearer ${jwtToken}` },
+      });
+
+      if (response.status === 200 && response.data) {
+        return {
+          success: true,
+          value: {
+            items: this.mapInvoices(response.data.items),
+            totalItems: response.data.totalItems,
+          },
+        };
+      } else {
+        return { success: false, error: response.data.title };
```

Comment:
If I am not wrong, in this place data can be null (I do not understand how TS does not throw an error), in that case reading `response.data.title` will throw an error, what do you think about returning all the response as the error? Or maybe adding a new Error as a wrapper (but I am not enough familiar with that approach).

```suggestion
        return { success: false, error: response };
```

Reply:
Done